### PR TITLE
rpc: implement websockets with github.com/gorilla/websocket 

### LINKED ISSUE #16846: Geth >= 1.8.3 can't deploy large contracts over websockets
#### System information

Geth version: `>= 1.8.3`
OS & Version: OSX
Commit hash : (if `develop`)

#### Expected behaviour
Should deploy contracts of any (legal) size over websockets. 

#### Actual behaviour
If the data sent has a length > 16416 the websocket connection fails and closes - all subsequent calls receive the error message:
```
Error: CONNECTION ERROR: Couldn't connect to node on WS.
``` 

#### Steps to reproduce the behaviour
Deploy a large contract over websockets to Geth 1.8.2 (Succeeds)
Deploy a large contract over websockets to Geth >= 1.8.3. (Errors)

Example of a contract that fails can be found at [ethpm/escape-truffle](https://github.com/ethpm/escape-truffle) where we are trying to run tests in CI using `geth --dev` but larger deployments like [this one](https://github.com/ethpm/escape-truffle/blob/master/contracts/ReleaseDB.sol) which consumes +/- 3 million gas trigger the error.  

This problem is also being tracked at web3 [here](https://github.com/ethereum/web3.js/issues/1548) because web3 crashes on the geth response.

@cz3kit discovered the data size limit mentioned above and may have more reproduction info.

### LINKED ISSUE #19798: rpc: WebSocket client treats pings as errors
Hello!

If the Go-Ethereum client connects to a server via WebSockets and uses `SubscribeNewHead`, then if that server sends a WebSocket ping message, the client treats this as an error and closes the connection. According to the WebSocket spec, either the client or server may send ping messages and the other endpoint should respond with a pong (see [MDN](https://developer.mozilla.org/en-US/docs/Web/API/WebSockets_API/Writing_WebSocket_servers#Pings_and_Pongs_The_Heartbeat_of_WebSockets) or the [WebSocket spec](https://tools.ietf.org/html/rfc6455#section-5.5.2)). Instead, if the server sends a ping then the connection is closed with an error.

A repro is as follows, which creates a go-ethereum client that subscribes to a mock server. The mock server responds to the `eth_subscribe` request, then sends a WebSocket ping after 12 seconds.
```go
package main

import (
	"context"
	"log"
	"sync"
	"time"

	"net/http"

	"github.com/ethereum/go-ethereum/core/types"
	"github.com/ethereum/go-ethereum/ethclient"
	"github.com/gorilla/websocket"
)

var waitGroup = &sync.WaitGroup{}
var upgrader = websocket.Upgrader{
	CheckOrigin: checkOrigin,
}

func main() {
	waitGroup.Add(2)
	go runServer()
	// Give server a moment to start up.
	time.Sleep(1 * time.Second)
	go runClient()
	waitGroup.Wait()
}

func runServer() {
	defer waitGroup.Done()
	http.HandleFunc("/", handleJSONRPCRequest)
	http.ListenAndServe("localhost:8080", nil)
}

func checkOrigin(r *http.Request) bool {
	return true
}

func handleJSONRPCRequest(writer http.ResponseWriter, request *http.Request) {
	conn, err := upgrader.Upgrade(writer, request, nil)
	if err != nil {
		log.Println(err)
		return
	}
	defer conn.Close()
	go func() {
		time.Sleep(12 * time.Second)
		log.Println("Server - Sending Ping")
		conn.WriteMessage(websocket.PingMessage, []byte("ping"))
	}()
	for {
		_, message, err := conn.ReadMessage()
		if err != nil {
			log.Println("Server - Read error:", err)
			break
		}
		log.Printf("Server - Received: %s", message)
		outMessage := []byte(`{"jsonrpc": "2.0", "id": 1, "result": "0x9ce59a13059e417087c02d3236a0b1cc"}`)
		err = conn.WriteMessage(websocket.TextMessage, outMessage)
		if err != nil {
			log.Println("Server - Write error:", err)
			break
		}
		log.Printf("Server - Sent: %s", outMessage)
	}
}

func runClient() {
	defer waitGroup.Done()
	endpoint := "ws://localhost:8080"
	ethWS, err := ethclient.Dial(endpoint)
	if err != nil {
		log.Fatal(err)
	}
	resultChan := make(chan *types.Header)
	subscription, err := ethWS.SubscribeNewHead(context.Background(), resultChan)
	if err != nil {
		log.Fatal(err)
	}
	log.Println("Client - Subscription successful, waiting for data ...")
	for {
		select {
		case sErr := <-subscription.Err():
			log.Println("Client - Saw error: ", sErr)
		case header := <-resultChan:
			log.Printf("Client - New block arrived, block_number: %v \n", header.Number)
		}
	}
}
```
Running this produces the output:
```
~/go/src/geth-repro$ ./geth-repro
2019/07/06 13:06:54 Server - Received: {"jsonrpc":"2.0","id":1,"method":"eth_subscribe","params":["newHeads"]}
2019/07/06 13:06:54 Server - Sent: {"jsonrpc": "2.0", "id": 1, "result": "0x9ce59a13059e417087c02d3236a0b1cc"}
2019/07/06 13:06:54 Client - Subscription successful, waiting for data ...
2019/07/06 13:07:06 Server - Sending Ping
2019/07/06 13:07:06 Client - Saw error:  write tcp 127.0.0.1:63875->127.0.0.1:8080: i/o timeout
2019/07/06 13:07:06 Server - Read error: websocket: close 1006 (abnormal closure): unexpected EOF
```

### PR DESCRIPTION
**Pull request by fjl**
This change makes package rpc use the github.com/gorilla/websocket package
for WebSockets instead of golang.org/x/net/websocket. The new library is more robust
and supports all WebSocket features including continuation frames.

There are new tests for two issues with the previously-used library:

  - TestWebsocketClientPing checks handling of Ping frames.
  - TestWebsocketLargeCall checks whether the request size limit is
    applied correctly.

DialWebsocket's handling of the default Origin header has changed. It used to set the
local machine's hostname as the origin, but that's weird and won't work with any
Internet-facing service. The new behavior is creating connections without an Origin
header unless one is provided explicitly.

This change also raises the request size limit for HTTP and WebSocket connections
to 5 MB. Quite a few people have reported issues where geth fails to deploy contracts
when they're too large, and retesteth needs a larger limit to upload the pre-state of large
tests.

Fixes #19798
Fixes #16846
Updates #19640 
Updates #19001

### CODE REVIEW

File: rpc/websocket.go [hunk 1]
```diff
@@ -112,60 +82,46 @@ func wsHandshakeValidator(allowedOrigins []string) func(*websocket.Config, *http
...
-
 	log.Debug(fmt.Sprintf("Allowed origin(s) for WS RPC interface %v", origins.ToSlice()))
 
-	f := func(cfg *websocket.Config, req *http.Request) error {
+	f := func(req *http.Request) bool {
 		// Skip origin verification if no Origin header is present. The origin check
 		// is supposed to protect against browser based attacks. Browsers always set
 		// Origin. Non-browser software can put anything in origin and checking it doesn't
 		// provide additional security.
 		if _, ok := req.Header["Origin"]; !ok {
```
Comment (karalabe) [inline rpc/websocket.go]: We need to use `req.Header.Get` here, otherwise the code won't ignore the case.

Comment (fjl) [reply rpc/websocket.go]: That's actually not true. net/http canonicalizes header keys, i.e. the keys in the map are assumed to be in the form returned by [http.CanonicalHeaderKey](https://godoc.org/net/http#CanonicalHeaderKey).

Comment (fjl) [reply rpc/websocket.go]: Using `req.Header.Get` is not possible because it doesn't distinguish between absent and empty headers.

**Comment by Deviad** [comment]
I have connection issues via websocket with web3j. Is it related to these changes?

**Comment by karalabe** [comment]
Possibly. Please provide more information and a repro.

**Comment by Deviad** [comment]
I started having problems with Infura (same connection issue) and for this reason I gave the go-ethereum client a shot. 
Here is where I establish the connection:
- [Connection](https://github.com/Deviad/clarity/blob/feature/add_ethereum_basic_operations_add_websockets_WIP/clarity-transaction-dispatcher/src/main/java/com/clarity/transactiondispatcher/services/InfuraSocketConfiguration.java)
Here is where I use this service:
- [Controller] (https://github.com/Deviad/clarity/blob/feature/add_ethereum_basic_operations_add_websockets_WIP/clarity-transaction-dispatcher/src/main/java/com/clarity/transactiondispatcher/web/controller/EthereumAccountController.java)

I know I had to use Websocket in the controller as well with stomp probably, but before to complicate things and get into an unexplored territory I wanted to be sure to be able to establish a connection and get some kind of result.

**Comment by Deviad** [comment]
Even with a different approach I have no luck:
```
   @SneakyThrows
    @Bean("webSocketSubscription")
    void wsConnectNetty() {
        Map<String, Object> config = Stream.of(
                new AbstractMap.SimpleEntry<>("includeTransactions", true)
        ).collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));

        Map<String, Object> map = Stream.of(
                new AbstractMap.SimpleEntry<>("id", 1),
                new AbstractMap.SimpleEntry<>("method", "eth_subscribe"),
                new AbstractMap.SimpleEntry<>("params", new Object[]{"newHeads", config})
        ).collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));


        ObjectMapper objectMapper = new ObjectMapper();

        String json = objectMapper.writeValueAsString(map);

        Flux<String> input = Flux.generate(sink -> sink.next(json));

        ReactorNettyWebSocketClient client = new ReactorNettyWebSocketClient();
        EmitterProcessor<String> output = EmitterProcessor.create();

        Mono<Void> sessionMono = client.execute(URI.create(web3jService.getInfuraWsEndpoint()), session -> session.send(input.map(session::textMessage))
                .thenMany(session.receive().map(WebSocketMessage::getPayloadAsText).doOnNext(log::info).subscribeWith(output).then()).then());

        output.doOnSubscribe(s -> {
            log.info("TEST >> {}", s.toString());
            sessionMono.subscribe();
        });
    }
```

**Pull request merged**